File: examples/latest/navigate/ios/Navigation/Navigation/NavigationExample.swift
/*
 * Copyright (C) 2019-2025 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

import AVFoundation
import heresdk
import SwiftUI

// Shows how to start and stop turn-by-turn navigation on a car route.
// By default, tracking mode is enabled. When navigation is stopped, tracking mode is enabled again.
// The preferred device language determines the language for voice notifications used for TTS.
// (Make sure to set language + region in device settings.)
class NavigationExample : DynamicRoutingDelegate, MessageDelegate {

    private let visualNavigator: VisualNavigator
    private let dynamicRoutingEngine: DynamicRoutingEngine
    private let herePositioningProvider: HEREPositioningProvider
    private let herePositioningSimulator: HEREPositioningSimulator
    private let routePrefetcher: RoutePrefetcher
    private let polygonPrefetcher: PolygonPrefetcher
    private let navigationHandler: NavigationHandler
    private let routeCalculator: RouteCalculator
    private let electronicHorizonHandler: ElectronicHorizonHandler
    var messageDelegate: MessageDelegate?

    init(mapView: MapView, routeCalculator:RouteCalculator) {
        do {
            // Without a route set, this starts tracking mode.
            try visualNavigator = VisualNavigator()
        } catch let engineInstantiationError {
            fatalError("Failed to initialize VisualNavigator. Cause: \(engineInstantiationError)")
        }

        self.routeCalculator = routeCalculator

        // By default, enable auto-zoom during guidance.
        visualNavigator.cameraBehavior = DynamicCameraBehavior()

        // By default, the MapView renders at 60 frames per second (fps).
        // When turn-by-turn navigation is enabled via the VisualNavigator,
        // the frame rate is reduced to 30 fps. This value can be customized;
        // for example, it is set to 60 fps below.
        visualNavigator.guidanceFrameRate = 60
        
        visualNavigator.startRendering(mapView: mapView)

        // A class to receive real location events.
        herePositioningProvider = HEREPositioningProvider()
        // A class to receive simulated location events.
        herePositioningSimulator = HEREPositioningSimulator()

        // The RoutePrefetcher downloads map data in advance into the map cache.
        // This is not mandatory, but can help to improve the guidance experience.
        routePrefetcher = RoutePrefetcher(SDKNativeEngine.sharedInstance!)
        polygonPrefetcher = PolygonPrefetcher(SDKNativeEngine.sharedInstance!)

        // An engine to find better routes during guidance.
        dynamicRoutingEngine = NavigationExample.createDynamicRoutingEngine()

        // A class to handle various kinds of guidance events.

        electronicHorizonHandler = ElectronicHorizonHandler()
        
        navigationHandler = NavigationHandler(
            visualNavigator, dynamicRoutingEngine, electronicHorizonHandler, routeCalculator)
        navigationHandler.messageDelegate = self
    }

    func startLocationProvider() {
        // Set navigator as delegate to receive locations from HERE Positioning.
        herePositioningProvider.startLocating(locationDelegate: visualNavigator,
                                              // Choose a suitable accuracy for the tbt navigation use case.
                                              accuracy: .navigation)
    }
    
    private class MyPrefetchStatusListener: PrefetchStatusListener {
        private var messageDelegate: MessageDelegate?

        init(messageDelegate: MessageDelegate?) {
            self.messageDelegate = messageDelegate
        }

        func onProgress(percentage: Int32) {
            messageDelegate?.updateMessage("Prefetch progress: \(percentage)%")
        }

        func onComplete(error: MapLoaderError?) {
            if let error = error {
                messageDelegate?.updateMessage("Prefetch failed: \(error)")
            } else {
                messageDelegate?.updateMessage("Prefetch completed successfully")
            }
        }
    }
    
    private func prefetchMapData(currentGeoCoordinates: GeoCoordinates) {
        // Prefetches map data around the provided location with a radius of 12 km into the map cache.
        // For the best experience, prefetch() should be called as early as possible.
        // Note that prefetchAroundRouteOnIntervals starts 1 km before reaching the end of the current corridor of 10 km (see below). 
        // Hence, for the first part of the route it is recommended to use the PolygonPrefetcher to cover the start of the route.
        let geoCircle = GeoCircle(center: currentGeoCoordinates, radiusInMeters: 10000.0)
        
        polygonPrefetcher.prefetch(
                geoPolygon: GeoPolygon(geoCircle: geoCircle),
                callback: { MyPrefetchStatusListener(messageDelegate: self.messageDelegate) }()
            )

        // Prefetches map data within a fixed-length corridor of 10 km along the route that is currently set to the provided Navigator instance.
        // This happens continuously in discrete intervals to fetch new corridors as the user is progressing along the route.
        // If no route is set, no data will be prefetched.
        // Alternatively, it is also possible to prefetch an entire route in advance using prefetchGeoCorridor(...).
        routePrefetcher.prefetchAroundRouteOnIntervals(navigator: visualNavigator)
    }
    
    // Use this engine to periodically search for better routes during guidance, ie. when the traffic
    // situation changes.
    //
    // Note: This code initiates periodic calls to the HERE Routing backend. Depending on your contract,
    // each call may be charged separately. It is the application's responsibility to decide how
    // often this code should be executed.
    private class func createDynamicRoutingEngine() -> DynamicRoutingEngine {
        // Both, minTimeDifference and minTimeDifferencePercentage, will be checked:
        // When the poll interval is reached, the smaller difference will win.
        let minTimeDifferencePercentage = 0.1
        let minTimeDifferenceInSeconds: TimeInterval = 1
        // Below, we use 10 minutes. A common range is between 5 and 15 minutes.
        let pollIntervalInSeconds: TimeInterval = 10 * 60

        let dynamicRoutingOptions =
        DynamicRoutingEngineOptions(minTimeDifferencePercentage: minTimeDifferencePercentage,
                                    minTimeDifference: minTimeDifferenceInSeconds,
                                    pollInterval: pollIntervalInSeconds)

        do {
            // With the dynamic routing engine you can poll the HERE backend services to search for routes with less traffic.
            // This can happen during guidance - or you can periodically update a route that is shown in a route planner.
            //
            // Make sure to call dynamicRoutingEngine.updateCurrentLocation(...) to trigger execution. If this is not called,
            // no events will be delivered even if the next poll interval has been reached.
            return try DynamicRoutingEngine(options: dynamicRoutingOptions)
        } catch let engineInstantiationError {
            fatalError("Failed to initialize DynamicRoutingEngine. Cause: \(engineInstantiationError)")
        }
    }

    // Conform to the DynamicRoutingDelegate.
    // Notifies on traffic-optimized routes that are considered better than the current route.
    func onBetterRouteFound(newRoute: Route,
                            etaDifferenceInSeconds: Int32,
                            distanceDifferenceInMeters: Int32) {
        print("DynamicRoutingEngine: Calculated a new route.")
        print("DynamicRoutingEngine: etaDifferenceInSeconds: \(etaDifferenceInSeconds).")
        print("DynamicRoutingEngine: distanceDifferenceInMeters: \(distanceDifferenceInMeters).")

        // An implementation needs to decide when to switch to the new route based
        // on above criteria.
    }

    // Conform to the DynamicRoutingDelegate.
    func onRoutingError(routingError: RoutingError) {
        print("Error while dynamically searching for a better route: \(routingError).")
    }

    func startNavigation(route: Route,
                         isSimulated: Bool) {
        let startGeoCoordinates = route.geometry.vertices[0]
        prefetchMapData(currentGeoCoordinates: startGeoCoordinates)

        // Switches to navigation mode when no route was set before, otherwise navigation mode is kept.
        visualNavigator.route = route

        if isSimulated {
            enableRoutePlayback(route: route)
            updateMessage("Starting simulated navigation.")
        } else {
            enableDevicePositioning()
            updateMessage("Starting navigation.")
        }

        startDynamicSearchForBetterRoutes(route)
        electronicHorizonHandler.start(route: route)
    }

    private func startDynamicSearchForBetterRoutes(_ route: Route) {
        do {
            // Note that the engine will be internally stopped, if it was started before.
            // Therefore, it is not necessary to stop the engine before starting it again.
            try dynamicRoutingEngine.start(route: route, delegate: self)
        } catch let instantiationError {
            fatalError("Start of DynamicRoutingEngine failed: \(instantiationError). Is the RouteHandle missing?")
        }
    }

    func stopNavigation() {
        // Switches to tracking mode when a route was set before, otherwise tracking mode is kept.
        // Without a route the navigator will only notify on the current map-matched location
        // including info such as speed and current street name.
        dynamicRoutingEngine.stop()
        electronicHorizonHandler.stop()
        routePrefetcher.stopPrefetchAroundRoute()
        visualNavigator.route = nil
        enableDevicePositioning()
        updateMessage("Tracking device's location.")
    }

    // Provides location updates based on the given route.
    func enableRoutePlayback(route: Route) {
        herePositioningProvider.stopLocating()
        herePositioningSimulator.startLocating(locationDelegate: visualNavigator, route: route)
    }

    // Provides location updates based on the device's GPS sensor.
    func enableDevicePositioning() {
        herePositioningSimulator.stopLocating()
        herePositioningProvider.startLocating(locationDelegate: visualNavigator,
                                              accuracy: .navigation)
    }

    func startCameraTracking() {
        visualNavigator.cameraBehavior = DynamicCameraBehavior()
    }

    func stopCameraTracking() {
        visualNavigator.cameraBehavior = nil
    }

    func getLastKnownLocation() -> Location? {
        return herePositioningProvider.getLastKnownLocation()
    }

    // Conform to MessageDelegate protocol.
    func updateMessage(_ message: String) {
        messageDelegate?.updateMessage(message)
    }
}

File: examples/latest/navigate/flutter/navigation_app/lib/NavigationExample.dart
/*
 * Copyright (C) 2019-2023 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

import 'package:flutter/foundation.dart';
import 'package:flutter/services.dart';
import 'package:here_sdk/core.dart';
import 'package:here_sdk/core.engine.dart';
import 'package:here_sdk/core.errors.dart';
import 'package:here_sdk/location.dart';
import 'package:here_sdk/mapview.dart';
import 'package:here_sdk/navigation.dart';
import 'package:here_sdk/prefetcher.dart';
import 'package:here_sdk/routing.dart' as HERE;
import 'package:here_sdk/routing.dart';
import 'package:here_sdk/trafficawarenavigation.dart';

import 'HEREPositioningProvider.dart';
import 'HEREPositioningSimulator.dart';

// Shows how to start and stop turn-by-turn navigation along a route.
class NavigationExample {
  final HereMapController _hereMapController;
  late VisualNavigator _visualNavigator;
  HEREPositioningSimulator _locationSimulationProvider;
  HEREPositioningProvider _herePositioningProvider;
  late DynamicRoutingEngine _dynamicRoutingEngine;
  final ValueChanged<String> _updateMessageState;
  RoutePrefetcher _routePrefetcher;

  NavigationExample(HereMapController hereMapController, ValueChanged<String> updateMessageState)
      : _hereMapController = hereMapController,
        _updateMessageState = updateMessageState,
        // For easy testing, this location provider simulates location events along a route.
        // You can use HERE positioning to feed real locations, see the "Positioning"-section in
        // our Developer's Guide for an example
        _locationSimulationProvider = HEREPositioningSimulator(),
        // Access the device's GPS sensor and other data.
        _herePositioningProvider = HEREPositioningProvider(),
        // The RoutePrefetcher downloads map data in advance into the map cache.
        // This is not mandatory, but can help to improve the guidance experience.
        _routePrefetcher = RoutePrefetcher(SDKNativeEngine.sharedInstance!) {
    try {
      _visualNavigator = VisualNavigator();
    } on InstantiationException {
      throw Exception("Initialization of VisualNavigator failed.");
    }

    // Enable auto-zoom during guidance.
    _visualNavigator.cameraBehavior = DynamicCameraBehavior();

    // This enables a navigation view including a rendered navigation arrow.
    _visualNavigator.startRendering(_hereMapController);

    // Set navigator as delegate to receive locations from HERE Positioning.
    // Choose a suitable accuracy for the tbt navigation use case.
    _herePositioningProvider.startLocating(_visualNavigator, LocationAccuracy.navigation);

    // An engine to find better routes during guidance.
    _createDynamicRoutingEngine();
  }

  void prefetchMapData(GeoCoordinates currentGeoCoordinates) {
    // Prefetches map data around the provided location with a radius of 2 km into the map cache.
    // For the best experience, prefetchAroundLocation() should be called as early as possible.
    _routePrefetcher.prefetchAroundLocation(currentGeoCoordinates);
    // Prefetches map data within a corridor along the route that is currently set to the provided Navigator instance.
    // This happens continuously in discrete intervals.
    // If no route is set, no data will be prefetched.
    _routePrefetcher.prefetchAroundRouteOnIntervals(_visualNavigator);
  }

  // Use this engine to periodically search for better routes during guidance, ie. when the traffic
  // situation changes.
  void _createDynamicRoutingEngine() {
    var dynamicRoutingOptions = DynamicRoutingEngineOptions();
    // Both, minTimeDifference and minTimeDifferencePercentage, will be checked:
    // When the poll interval is reached, the smaller difference will win.
    dynamicRoutingOptions.minTimeDifference = Duration(seconds: 1);
    dynamicRoutingOptions.minTimeDifferencePercentage = 0.1;
    dynamicRoutingOptions.pollInterval = Duration(minutes: 5);

    try {
      // With the dynamic routing engine you can poll the HERE backend services to search for routes with less traffic.
      // This can happen during guidance - or you can periodically update a route that is shown in a route planner.
      //
      // Make sure to call dynamicRoutingEngine.updateCurrentLocation(...) to trigger execution. If this is not called,
      // no events will be delivered even if the next poll interval has been reached.
      _dynamicRoutingEngine = DynamicRoutingEngine(dynamicRoutingOptions);
    } on InstantiationException {
      throw Exception("Initialization of DynamicRoutingEngine failed.");
    }
  }

  Location? getLastKnownLocation() {
    return _herePositioningProvider.getLastKnownLocation();
  }

  void startNavigationSimulation(HERE.Route route) {
    // Set the route to follow.
    _visualNavigator.route = route;

    // Stop in case it was started before.
    _herePositioningProvider.stop();

    // Simulates location events based on the given route.
    // The navigator is set as listener to receive location updates.
    _locationSimulationProvider.startLocating(route, _visualNavigator);

    _startDynamicSearchForBetterRoutes(route);
  }

  void startNavigation(HERE.Route route) {
    GeoCoordinates startGeoCoordinates = route.geometry.vertices[0];
    prefetchMapData(startGeoCoordinates);

    // Set the route to follow.
    _visualNavigator.route = route;

    // Stop in case it was started before.
    _locationSimulationProvider.stop();

    // Access the device's GPS sensor and other data.
    // The navigator is set as listener to receive location updates.
    _herePositioningProvider.startLocating(_visualNavigator, LocationAccuracy.navigation);

    _startDynamicSearchForBetterRoutes(route);
  }

  void _startDynamicSearchForBetterRoutes(HERE.Route route) {
    try {
      // Note that the engine will be internally stopped, if it was started before.
      // Therefore, it is not necessary to stop the engine before starting it again.
      _dynamicRoutingEngine.start(
          route,
          // Notifies on traffic-optimized routes that are considered better than the current route.
          DynamicRoutingListener((Route newRoute, int etaDifferenceInSeconds, int distanceDifferenceInMeters) {
            _updateMessageState("DynamicRoutingEngine: Calculated a new route");
            print("DynamicRoutingEngine: etaDifferenceInSeconds: $etaDifferenceInSeconds.");
            print("DynamicRoutingEngine: distanceDifferenceInMeters: $distanceDifferenceInMeters.");

            // An implementation needs to decide when to switch to the new route based
            // on above criteria.
          }, (RoutingError routingError) {
            final error = routingError.toString();
            _updateMessageState("Error while dynamically searching for a better route: $error");
          }));
    } on DynamicRoutingEngineStartException {
      throw Exception("Start of DynamicRoutingEngine failed. Is the RouteHandle missing?");
    }
  }

  void followCurrentCarPosition(bool isFollowing) {
    if (isFollowing) {
      _visualNavigator.cameraBehavior = DynamicCameraBehavior();
    } else {
      _visualNavigator.cameraBehavior = null;
    }
  }

  void stopNavigation() {
    // Stop in case it was started before.
    _locationSimulationProvider.stop();
    _dynamicRoutingEngine.stop();
    _routePrefetcher.stopPrefetchAroundRoute();
    _startTracking();
    _updateMessageState("Tracking device's location.");
  }

  // Starts tracking the device's location using HERE Positioning.
  void _startTracking() {
    // Leaves guidance (if it was running) and enables tracking mode. The camera may optionally follow, see toggleTracking().
    _visualNavigator.route = null;
    _herePositioningProvider.startLocating(_visualNavigator, LocationAccuracy.navigation);
  }

  void detach() {
    // It is recommended to stop rendering before leaving the app.
    // This also removes the current location marker.
    _visualNavigator.stopRendering();

    // Stop LocationSimulator and DynamicRoutingEngine in case they were started before.
    _locationSimulationProvider.stop();
    _dynamicRoutingEngine.stop();

    // It is recommended to stop the LocationEngine before leaving the app.
    _herePositioningProvider.stop();
  }
}
